### PR TITLE
Unseal

### PR DESCRIPTION
**Pull request by magik6k**
### tldr

This PR adds an unseal pipeline along with a simple tagging system on sectors which allows SPs to specify the desired sector unseal state.

#### Sector target-unseal mark

Sectors now have a `target_unseal_state` marker which has 3 states:
* `none` (default) - don't do anything relating to the unsealed file. If it's there don't remove, if it's not there don't unseal
* `false` - mark for GC when it appears in storage (only removes after the GC mark is separately approved)
* `true` - currently, when the value flips from other state to true AND the unsealed file isn't present in storage an entry for the sector will be added to the unseal pipeline
  * Eventually there will be a separate mechanism monitoring storage for missing unsealed files that might be able to perform recovery, but for now to avoid triggering accidental unseals when some storage gets disconnected temporary from the cluster, unseals only get scheduled on target state change.

#### Unseal Pipeline

The unseal pipeline is vastly more robust than unseal requests in lotus-miner ever were, and at the same time much simpler. There are only two tasks:
* `SDRKeyRegen` - creates sector key on an SDR worker
  * This key is stored in scratch space
  * Only the key (last layer) is kept on-disk after the compute is done
* `UnsealDecode` - gets (potentially remote) streams of the sector key and sealed data, and outputs unsealed data straight to long-term-storage
  * This handles both snap and sdr unseal. Both decode processes appear to be fast enough to run even on quite under-powered storage machines
  * After this step is done the sector key is removed from scratch space

#### Tools

Some tools added in this PR:
* `curio unseal info` - outputs most relevant information about unsealed sector data
2024-09-05-015827_640x234_scrot

* `curio unseal list-sectors` - very basic aggregate output dumping the unseal pipeline state + target unseal states, output is not pretty but enough for visibility
2024-09-05-015744_1550x616_scrot

* `curio unseal set-target-state [spid] [sector] [true/false/none]` - setting target unseal state for sectors
* `curio unseal check [spid] [sector]` - run a CommP on stored unsealed file to verify its integrity
2024-09-05-020427_728x113_scrot

### todo

* [x] Generate Key
* [x] Decode
  * [x] BONUS: Snap
* [x] Pipeline
* [x] CommK GC
* [x] Wire into curio run
* [x] Default build works
* [x] Command to mark sectors for unseal
* [x] Shared SDR max counter
* [x] GC awareness for target-no-unsealed
* [x] MetadataGC for Unseal pipeline
* [x] Test run
* [x] Command to verify unsealed sector data
* [x] Change commp check config to mach market commp

### CODE REVIEW

File: lib/ffi/cunative/decode_snap.go [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+	el, err := c.Element.SetString(s)
+	if err != nil {
+		return nil, err
+	}
+
+	c.Element = el
+	return c, nil
+}
+
+var _ poseidon.Element[*CursedPoseidonGenRandomnessElement] = &CursedPoseidonGenRandomnessElement{}
```
Comment (snadrus) [inline lib/ffi/cunative/decode_snap.go:525]: Cursed?

File: lib/ffi/sdr_funcs.go [hunk 2]
```diff
@@ -143,43 +143,67 @@ func (l *storageProvider) AcquireSector(ctx context.Context, taskID *harmonytask
...
 	}
 
-	replicaID, err := sector.ProofType.ReplicaId(sector.ID.Miner, sector.ID.Number, ticket, commp)
+	replicaID, err := sector.ProofType.ReplicaId(sector.ID.Miner, sector.ID.Number, ticket, commd)
 	if err != nil {
 		return xerrors.Errorf("computing replica id: %w", err)
 	}
 
+	intoPath := storiface.PathByType(paths, into)
+	intoTemp := intoPath + ".tmp"
```
Comment (snadrus) [inline lib/ffi/sdr_funcs.go:165]: storiface.TempSuffix

File: lib/proof/datacid.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+const commPBufPad = abi.PaddedPieceSize(8 << 20)
+const CommPBuf = abi.UnpaddedPieceSize(commPBufPad - (commPBufPad / 128)) // can't use .Unpadded() for const
+
+type ciderr struct {
+	c   cid.Cid
+	err error
+}
+
+/*
+DataCidWriter is used as follows:
```
Comment (snadrus) [inline lib/proof/datacid.go:32]: Add a bit as to why I want to use this:
This allows streaming reads to compute the PieceCID, PayloadSize, and padded piece size

File: tasks/gc/pipeline_meta_gc.go [hunk 4]
```diff
@@ -149,5 +152,24 @@ func (s *PipelineGC) cleanupUpgrade() error {
 	return nil
 }
 
+func (s *PipelineGC) cleanupUnseal() error {
+	// Remove sectors_unseal_pipeline entries where:
```
Comment (snadrus) [inline tasks/gc/pipeline_meta_gc.go:156]: This cleans up unseal records that don't have any pending steps?

File: tasks/scrub/task_scrub_commd.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+func (c *ScrubCommDTask) schedule(ctx context.Context, taskFunc harmonytask.AddTaskFunc) error {
+	taskFunc(func(id harmonytask.TaskID, tx *harmonydb.Tx) (shouldCommit bool, seriousError error) {
+		var checks []struct {
+			CheckID      int64 `db:"check_id"`
+			SpID         int64 `db:"sp_id"`
+			SectorNumber int64 `db:"sector_number"`
+		}
+
+		err := c.db.Select(ctx, &checks, `
```
Comment (snadrus) [inline tasks/scrub/task_scrub_commd.go:133]: tx.Select

File: tasks/scrub/task_scrub_commd.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+		`)
+		if err != nil {
+			return false, xerrors.Errorf("getting tasks: %w", err)
+		}
+
+		if len(checks) == 0 {
+			return false, nil
+		}
+
+		// pick at random in case there are a bunch of schedules across the cluster
```
Comment (snadrus) [inline tasks/scrub/task_scrub_commd.go:146]: oldest?

File: tasks/unseal/task_unseal_decode.go [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+			return false, xerrors.Errorf("DecodeSnap: %w", err)
+		}
+	} else {
+		err = t.sc.DecodeSDR(ctx, taskID, sref)
+		if err != nil {
+			return false, xerrors.Errorf("DecodeSDR: %w", err)
+		}
+	}
+
+	_, err = t.db.Exec(ctx, `UPDATE sectors_unseal_pipeline SET after_decode_sector = TRUE, task_id_decode_sector = NULL WHERE task_id_decode_sector = $1`, taskID)
```
Comment (snadrus) [inline tasks/unseal/task_unseal_decode.go:144]: Where do we delete the key file?

**Review by snadrus — APPROVED** [review]

**Review by LexLuthr — APPROVED** [review]
Looks good with some small nits.

File: tasks/gc/pipeline_meta_gc.go (hunk 4 above)
Comment (magik6k) [reply tasks/gc/pipeline_meta_gc.go:156]: Correct, entries that have had all the work done already

File: tasks/unseal/task_unseal_decode.go (hunk 7 above)
Comment (magik6k) [reply tasks/unseal/task_unseal_decode.go:144]: Bottom of SealCalls.decodeCommon, will add a comment here

File: cmd/curio/unseal.go [hunk 8]
```diff
@@ -0,0 +1,561 @@
...
+			return xerrors.Errorf("error iterating rows: %w", err)
+		}
+
+		fmt.Println("Data exported successfully.")
+		return nil
+	},
+}
+
+var setTargetUnsealStateCmd = &cli.Command{
+	Name:      "set-target-state",
```
[the remark this answers is not in the document]
Comment (magik6k) [reply cmd/curio/unseal.go:394]: The command is already `curio unseal set-target-state`, no need to say the same word twice imo.

File: cmd/curio/unseal.go [hunk 9]
```diff
@@ -0,0 +1,561 @@
...
+					fmt.Printf("Check task failed %s\n", color.RedString("✘"))
+					fmt.Printf("Error message: %s\n", message.String)
+					fmt.Printf("Actual unsealed CID:   %s\n", actualUnsealedCID.String)
+				}
+				return nil
+			}
+
+			// Update progress indicator
+			dots = (dots + 1) % 4
+			_, _ = fmt.Fprintf(os.Stderr, "Check task still in progress%s    \r", "."+strings.Repeat(".", dots))
```
[the remark this answers is not in the document]
Comment (magik6k) [reply cmd/curio/unseal.go:556]: In general dynamic UI output shouldn't go to stdout, eventually someone somewhere will pipe this to a file and try to parse the output, and outputs with \r are very confusing, so putting that output in stderr is generally a better practice

File: cmd/curio/unseal.go [hunk 10]
```diff
@@ -0,0 +1,561 @@
...
+}
+
+func filterPred[T any](pred func(T) bool) func(T, int) bool {
+	return func(item T, _ int) bool {
+		return pred(item)
+	}
+}
+
+var listUnsealPipelineCmd = &cli.Command{
+	Name:  "list-sectors",
```
[the remark this answers is not in the document]
Comment (magik6k) [reply cmd/curio/unseal.go:275]: The command is  `curio unseal list-sectors`, which is imho good enough for a command which gives you a basic listing of sectors with their unseal-related info

File: lib/ffi/cunative/decode_snap.go (hunk 1 above)
Comment (magik6k) [reply lib/ffi/cunative/decode_snap.go]: A little.

Gave this a slightly better name

File: cmd/curio/unseal.go [hunk 11]
```diff
@@ -0,0 +1,561 @@
...
+		fmt.Println("Data exported successfully.")
+		return nil
+	},
+}
+
+var setTargetUnsealStateCmd = &cli.Command{
+	Name:      "set-target-state",
+	Usage:     "Set the target unseal state for a sector",
+	ArgsUsage: "<sp-id> <sector-number> <target-state>",
+	Description: `Set the target unseal state for a specific sector.
```
[the remark this answers is not in the document]
Comment (magik6k) [reply cmd/curio/unseal.go:397]: Expanded the description

File: harmony/harmonytask/harmonytask.go [hunk 12]
```diff
@@ -22,8 +22,9 @@ var FOLLOW_FREQUENCY = 1 * time.Minute          // Check for work to follow this
 
 type TaskTypeDetails struct {
 	// Max returns how many tasks this machine can run of this type.
-	// Zero (default) or less means unrestricted.
-	Max int
+	// Nil (default)/Zero or less means unrestricted.
+	// Counters can either be independent when created with Max, or shared between tasks with SharedMax.Make()
+	Max *MaxCounter
```
[the remark this answers is not in the document]
Comment (magik6k) [reply harmony/harmonytask/harmonytask.go]: Done
